api/src/main/java/org/openmrs/module/reporting/common/SqlIterator.java

```diff
@@ -0,0 +1,91 @@
+package org.openmrs.module.reporting.common;
+
+import org.openmrs.api.APIException;
+import org.openmrs.module.reporting.dataset.DataSetColumn;
+import org.openmrs.module.reporting.dataset.DataSetRow;
+
+import java.sql.Connection;
+import java.sql.ResultSet;
+import java.sql.ResultSetMetaData;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.NoSuchElementException;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class SqlIterator implements Iterator<DataSetRow> {
+    private ResultSet resultSet;
+    private List<DataSetColumn> columns;
+    private Statement statement;
+    private Connection connection;
+    private static final Logger log = LoggerFactory.getLogger(SqlIterator.class);
+
+
+    public SqlIterator(ResultSetMetaData metadata, ResultSet resultSet, Connection connection, Statement statement) throws SQLException {
+        this.resultSet = resultSet;
+        this.connection = connection;
+        this.statement = statement;
+        this.createDataSetColumns(metadata);
+    }
+
+    @Override
+    public boolean hasNext() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public DataSetRow next() throws NoSuchElementException {
+        try {
+            if (resultSet.next()) {
+                return createDataSetRow();
+            } else {
+                closeConnection();
```

Comment:
My first instinct is that this closeConnection() statement (and the entire closeConnection method below) should not be in this class.  This SqlIterator instance does not create the connection - it is passed an existing connection in it's constructor - so I wouldn't think it should take any responsibility over closing it.  And if it were to do this (and I don't think it should), it should likely do it consistently - i.e. right now you are not closing the connection if an exception occurs in the catch block below.  So a client will end up in various unpredicatable states with their connection depending on what happens inside this class.  So I think we should remove the connection closing from this.

Reply:
Regarding the ownership of the connection, I agree with you. The reason that why is closed here is because the class `SqlRunner`  where this Iterator is created it does not know the `IterableSqlDataSetEvaluator` that create the connection. 
I can make sure that `IterableSqlDataSetEvaluator` is the one responsible to close the connection but for that the methods inside `SqlRunner` should receive the `evaluator` instead of the connection, and I don't feel comfortable messing with the other methods.  

Also in the end the `SqlIterator` is the one o knows when the connection is no longer required. 

Note: I added the close connection to the exception.